File: Stack/Valid Parentheses/ValidParentheses.cpp
class Solution
{
public:
    bool isValid(string s)
    {
        stack<char> st;
        for (auto i : s)
        {
            if (st.empty() && (i == ')' || i == ']' || i == '}'))
            {
                return false;
            }
            else if (i == '(' || i == '[' || i == '{')
            {
                st.push(i);
            }
            else if (!st.empty() && st.top() == '[' && i == ']')
            {
                st.pop();
            }
            else if (!st.empty() && st.top() == '(' && i == ')')
            {
                st.pop();
            }
            else if (!st.empty() && st.top() == '{' && i == '}')
            {
                st.pop();
            }
            else
                return false;
        }
        if (st.size() == 0)
            return true;
        return false;
    }
};

File: Stack/Valid Parentheses/ValidParentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if (len(s) % 2 != 0):
            return False

        pairs = {
            "}": "{",
            "]": "[",
            ")": "("
        }
        stack = []
        
        for i in s:
            if i in pairs:
                if len(stack) > 0 and pairs[i] == stack[-1]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(i)

        return len(stack) == 0
